library/core/src/num/nonzero.rs

```diff
@@ -1543,33 +1543,33 @@ macro_rules! nonzero_integer_signedness_dependent_methods {
         #[rustc_const_unstable(feature = "isqrt", issue = "116226")]
         #[must_use = "this returns the result of the operation, \
                       without modifying the original"]
-        #[inline]
         pub const fn isqrt(self) -> Self {
```

Comment:
OK, I've changed it to call the `uint_macros` function:

<details>
<summary>The new function code</summary>

```rust
        #[inline(always)]
        pub const fn isqrt(self) -> Self {
            let result = self.get().isqrt();

            // Inform the optimizer what the range of outputs is. The upper
            // bound was already handled by the `$UInt::isqrt` call just above.
            //
            // SAFETY: Integer square root is a monotonically nondecreasing
            // function, which means that increasing the input will never cause
            // the output to decrease. Thus, since the input for nonzero
            // unsigned integers has a lower bound of 1, the lower bound of the
            // result will be sqrt(1), which is 1.
            unsafe {
                hint::assert_unchecked(result >= 1);
            }

            // SAFETY: As explained above, the minimum integer square root is
            // 1, so it can't be 0.
            unsafe { Self::new_unchecked(result) }
        }
```
</details>

```

Reply:
One thing I've found is that `NonZero<usize>` calls won't inline `core::num::nonzero::NonZero<usize>::isqrt`, but `NonZero<u64>` will inline far past `core::num::nonzero::NonZero<u64>::isqrt`. I'll try to investigate why.

<details>
<summary>Code generated by `u64` vs `usize` function argument</summary>

```rust
#[inline(never)]
pub fn isqrt_test_5(n: NonZero<u64>) -> u32 {
    u32::try_from(n.isqrt().get()).unwrap()
}
```

```
testing::isqrt_test_5:
Lfunc_begin4:
        .cfi_startproc
        stp x29, x30, [sp, #-16]!
        .cfi_def_cfa_offset 16
        mov x29, sp
        .cfi_def_cfa w29, 16
        .cfi_offset w30, -8
        .cfi_offset w29, -16
        bl core::num::int_sqrt::u64
        .cfi_def_cfa wsp, 16
        ldp x29, x30, [sp], #16
        .cfi_def_cfa_offset 0
        .cfi_restore w30
        .cfi_restore w29
        ret
```

vs

```rust
#[inline(never)]
pub fn isqrt_test_5(n: NonZero<usize>) -> u32 {
    u32::try_from(n.isqrt().get()).unwrap()
}
```

```
testing::isqrt_test_5:
Lfunc_begin6:
	.cfi_startproc
	sub sp, sp, #32
	.cfi_def_cfa_offset 32
	stp x29, x30, [sp, #16]
	add x29, sp, #16
	.cfi_def_cfa w29, 16
	.cfi_offset w30, -8
	.cfi_offset w29, -16
	.cfi_remember_state
	bl core::num::nonzero::NonZero<usize>::isqrt
	lsr x8, x0, #32
	cbnz x8, LBB6_2
	.cfi_def_cfa wsp, 32
	ldp x29, x30, [sp, #16]
	add sp, sp, #32
	.cfi_def_cfa_offset 0
	.cfi_restore w30
	.cfi_restore w29
	ret
	.cfi_restore_state
Lloh14:
	adrp x0, l___unnamed_9@PAGE
Lloh15:
	add x0, x0, l___unnamed_9@PAGEOFF
Lloh16:
	adrp x3, l___unnamed_1@PAGE
Lloh17:
	add x3, x3, l___unnamed_1@PAGEOFF
Lloh18:
	adrp x4, l___unnamed_10@PAGE
Lloh19:
	add x4, x4, l___unnamed_10@PAGEOFF
	sub x2, x29, #1
	mov w1, #43
	bl core::result::unwrap_failed
```
</details>

Reply:
With `#[inline(always)]` on this `isqrt` function, it inlines all the way to the background function that does the actual square root (`core::num::int_sqrt::u*`) and reduces the code size as well. Without it, it won't even inline past `core::num::nonzero::NonZero<u*>::isqrt`:

<details>
<summary>With and without `#[inline(always)]`</summary>
The function:

```rust
#[inline(never)]
pub fn isqrt_test_5(n: NonZero<usize>) -> u32 {
    u32::try_from(n.isqrt().get()).unwrap()
}
```

With `#[inline(always)]`:

```
testing::isqrt_test_5:
Lfunc_begin4:
        .cfi_startproc
        stp x29, x30, [sp, #-16]!
        .cfi_def_cfa_offset 16
        mov x29, sp
        .cfi_def_cfa w29, 16
        .cfi_offset w30, -8
        .cfi_offset w29, -16
        bl core::num::int_sqrt::u64
        .cfi_def_cfa wsp, 16
        ldp x29, x30, [sp], #16
        .cfi_def_cfa_offset 0
        .cfi_restore w30
        .cfi_restore w29
        ret
```

Without any `inline` attribute:

```
testing::isqrt_test_5:
Lfunc_begin6:
        .cfi_startproc
        sub sp, sp, #32
        .cfi_def_cfa_offset 32
        stp x29, x30, [sp, #16]
        add x29, sp, #16
        .cfi_def_cfa w29, 16
        .cfi_offset w30, -8
        .cfi_offset w29, -16
        .cfi_remember_state
        bl core::num::nonzero::NonZero<usize>::isqrt
        lsr x8, x0, #32
        cbnz x8, LBB6_2
        .cfi_def_cfa wsp, 32
        ldp x29, x30, [sp, #16]
        add sp, sp, #32
        .cfi_def_cfa_offset 0
        .cfi_restore w30
        .cfi_restore w29
        ret
        .cfi_restore_state
Lloh14:
        adrp x0, l___unnamed_9@PAGE
Lloh15:
        add x0, x0, l___unnamed_9@PAGEOFF
Lloh16:
        adrp x3, l___unnamed_1@PAGE
Lloh17:
        add x3, x3, l___unnamed_1@PAGEOFF
Lloh18:
        adrp x4, l___unnamed_10@PAGE
Lloh19:
        add x4, x4, l___unnamed_10@PAGEOFF
        sub x2, x29, #1
        mov w1, #43
        bl core::result::unwrap_failed
        .loh AdrpAdd        Lloh18, Lloh19
        .loh AdrpAdd        Lloh16, Lloh17
        .loh AdrpAdd        Lloh14, Lloh15
```
</details>

Reply:
This looks better!

So the wrapper always gets inlined but the routine itself can outline? That is good, as long LLVM has a point to not allow inlining. I requested comments in a few places to make this obvious.

You should try with only `#[inline]` (not always) to see if you get the same results, and prefer that if so. It just allows LLVM pick slightly different points for outlining when optimizing for size.

Reply:
OK, I've changed the public `checked_isqrt`/`isqrt` methods and the internal functions to `#[inline]` and, for 64-bit inputs, it inlines everything except for a call to `core::num::int_sqrt::u64`, which is good.

Reply:
I think you can do this

<img width="886" alt="image" src="https://github.com/user-attachments/assets/59a5e697-4916-4e55-8394-332ee5d3ec27">

But doesn't matter, there shouldn't be much more to do here. Thanks for the help with the llvm stuff :)